db/migrate/20240116180000_fix_unread_comments_inconsistencies.rb

```diff
@@ -0,0 +1,59 @@
+# Fixes the unread_comments flag for all users. Unintended behavior was
+# introduced in pull request #515. Migration introduced in #587.
+# Behavior fixed in #585.
+#
+# This migration is generally *not* idempotent since users might have interacted
+# with the website since the migration was run and thus they will probably have
+# different unread comments flags as the ones at the time of the migration.
+#
+# This migration is not reversible as we don't store the previous state of
+# the unread_comments flag.
+class FixUnreadCommentsInconsistencies < ActiveRecord::Migration[7.0]
+  def up
+    num_fixed_users = 0
+
+    User.find_each do |user|
+      had_user_unread_comments = user.unread_comments # boolean
+      has_user_unread_comments = fix_unread_comments_flag(user)
+
+      has_flag_changed = had_user_unread_comments != has_user_unread_comments
+      user.update(unread_comments: has_user_unread_comments) if has_flag_changed
+      num_fixed_users += 1 if has_flag_changed
+    end
+
+    Rails.logger.debug { "Ran through #{User.count} users (unread comments flag)" }
+    Rails.logger.debug { "Fixed #{num_fixed_users} users (unread comments flag)" }
+  end
+
+  # Fixes the unread_comments flag for a given user.
+  # Returns whether the user has unread comments (after a possible change).
+  def fix_unread_comments_flag(user)
+    # Check for unread comments -- directly via Reader
+    readers = Reader.where(user: user)
+    readers.each do |reader|
+      thread = Commontator::Thread.find_by(id: reader.thread_id)
+      next if thread.nil?
+
+      latest_thread_comment_by_any_user = thread.comments.max_by(&:created_at)
+      next if latest_thread_comment_by_any_user.blank?
+
+      latest_thread_comment_time = latest_thread_comment_by_any_user.created_at
+      has_user_unread_comments = reader.updated_at < latest_thread_comment_time
+
+      return true if has_user_unread_comments
+    end
+
+    # User might still have unread comments but no related Reader objects
+    # -> Check for unread comments -- via Media
+    # TODO: rename media_latest_comments to subscribed_media_with_latest_comments_not_by_creator
+    # see change in PR #585
+    unseen_media = user.media_latest_comments.select! do |m|
```

Comment:
Isn't #587 going to be merged into dev after #585? Because #585 replaces the `media_latest_comments`
method of `User` by a method called `subscribed_media_with_latest_comments_not_by_creator`. The migration would then throw an error because this method no longer exists:

```
Migrating to FixUnreadCommentsInconsistencies (20240116180000)
== 20240116180000 FixUnreadCommentsInconsistencies: migrating =================
  TRANSACTION (0.2ms)  BEGIN
  User Load (0.5ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1000]]
  Reader Load (0.3ms)  SELECT "readers".* FROM "readers" WHERE "readers"."user_id" = $1  [["user_id", 1]]
  TRANSACTION (0.2ms)  ROLLBACK
   (0.3ms)  SELECT pg_advisory_unlock(8267686635516738675)
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

undefined method `media_latest_comments' for #<User id: 1, email: "admin@mampf.edu", created_at: "2020-07-24 15:07:48.807973000 +0200", updat...
```

Before I noticed that, I had checked the branch out from `dev`, and then the migration ran through, but created  new inconsistencies:
![Screenshot 2024-01-16 165217](https://github.com/MaMpf-HD/mampf/assets/28628554/13261ae6-6061-4e60-9d35-12152c9573d7)

My hope was then that if I merged `fix/unread-comments-db-inconsistencies` into `fix/unread-comments` and replaced `media_latest_comments` in the migration by `subscribed_media_with_latest_comments_not_by_creator` then the newly created inconsistencies would go away after migrating, but they did not.

Reply:
> Isn't https://github.com/MaMpf-HD/mampf/pull/587 going to be merged into dev after https://github.com/MaMpf-HD/mampf/pull/585?

Yes, that's true and which is why I wrote the respective "TODO" in the code as this has to be done after #585 is in `dev` (I'm going to merge #585 soon, but wanted to wait until this PR here is ready to go except for this method renaming).

---

> My hope was ...

The way you merged branches, it should have actually got rid of any inconsistencies. I'm keen to see the exact case in your local setup in the meeting.

Reply:
> The problem seems to be that you use select! instead of select

Thanks a lot, that was the problem. This made it to the hall of fame of dumb bugs I introduced, see [here](https://github.com/MaMpf-HD/mampf/wiki/Code#hall-of-fame-of-dumb-ruby-bugs-%EF%B8%8F).